From ec003017df30f5ac0125042673b7a4663458327f Mon Sep 17 00:00:00 2001
From: Fabio Baltieri <fabiobaltieri@google.com>
Date: Fri, 12 Jun 2026 19:30:49 +0100
Subject: [PATCH] drivers: led: add a Linux LED native_sim driver

Add a native_sim driver to control a Linux LED on the host using the
Zephyr LED APIs.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>
---
 boards/native/native_sim/doc/index.rst        |  20 ++
 drivers/led/CMakeLists.txt                    |   8 +
 drivers/led/Kconfig                           |   1 +
 drivers/led/Kconfig.linux-leds                |  13 ++
 drivers/led/linux_leds.c                      | 196 ++++++++++++++++++
 .../led/zephyr,native-linux-leds.yaml         |  37 ++++
 tests/drivers/build_all/led/app.overlay       |  12 ++
 7 files changed, 287 insertions(+)
 create mode 100644 drivers/led/Kconfig.linux-leds
 create mode 100644 drivers/led/linux_leds.c
 create mode 100644 dts/bindings/led/zephyr,native-linux-leds.yaml

diff --git a/boards/native/native_sim/doc/index.rst b/boards/native/native_sim/doc/index.rst
index ee26998f6f82..18a13c9fa26f 100644
--- a/boards/native/native_sim/doc/index.rst
+++ b/boards/native/native_sim/doc/index.rst
@@ -543,6 +543,26 @@ Here are more details on the peripherals that are currently provided with this b
 
   SocketCAN support can be enabled by using the :ref:`snippet-socketcan-native-sim`.
 
+**LED device**
+  Implements a Zephyr LED device backed by a Linux LED. You configure which
+  Linux LED to use by setting the DT ``path`` property to the name of the LED
+  as it shows in ``/sys/class/leds``.
+
+  .. code-block:: dts
+
+      leds {
+        compatible = "zephyr,native-linux-leds";
+
+        led0: led_0 {
+          /* /sys/class/leds/tpacpi::lid_logo_dot/brightness */
+          path = "tpacpi::lid_logo_dot";
+        };
+        led1: led_1 {
+          /* /sys/class/leds/tpacpi::power/brightness */
+          path = "tpacpi::power";
+        };
+      };
+
 .. _native_ptty_uart:
 
 PTY UART
diff --git a/drivers/led/CMakeLists.txt b/drivers/led/CMakeLists.txt
index 774e4efc0894..3e5a1750f823 100644
--- a/drivers/led/CMakeLists.txt
+++ b/drivers/led/CMakeLists.txt
@@ -33,3 +33,11 @@ zephyr_library_sources_ifdef(CONFIG_PCA9533 pca9533.c)
 zephyr_library_sources_ifdef(CONFIG_PCA9633 pca9633.c)
 zephyr_library_sources_ifdef(CONFIG_TLC59108 tlc59108.c)
 # zephyr-keep-sorted-stop
+
+if(CONFIG_NATIVE_LINUX_LEDS)
+  if(${CMAKE_HOST_SYSTEM_NAME} STREQUAL Linux)
+    zephyr_library_sources(linux_leds.c)
+  else()
+    message(FATAL_ERROR "CONFIG_NATIVE_LINUX_LEDS is only available on Linux")
+  endif()
+endif()
diff --git a/drivers/led/Kconfig b/drivers/led/Kconfig
index d076843acc55..972543628732 100644
--- a/drivers/led/Kconfig
+++ b/drivers/led/Kconfig
@@ -36,6 +36,7 @@ source "drivers/led/Kconfig.is31fl319x"
 source "drivers/led/Kconfig.is31fl3216a"
 source "drivers/led/Kconfig.is31fl3733"
 source "drivers/led/Kconfig.leds-group-multicolor"
+source "drivers/led/Kconfig.linux-leds"
 source "drivers/led/Kconfig.lp3943"
 source "drivers/led/Kconfig.lp50xx"
 source "drivers/led/Kconfig.lp5562"
diff --git a/drivers/led/Kconfig.linux-leds b/drivers/led/Kconfig.linux-leds
new file mode 100644
index 000000000000..597518c3e759
--- /dev/null
+++ b/drivers/led/Kconfig.linux-leds
@@ -0,0 +1,13 @@
+# Copyright 2026 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+config NATIVE_LINUX_LEDS
+	bool "Native Linux LEDs based LED device"
+	default y
+	depends on DT_HAS_ZEPHYR_NATIVE_LINUX_LEDS_ENABLED
+	depends on ARCH_POSIX
+	help
+	  Enable a native_sim LED driver backed by Linux sysfs LEDs.
+	  Each DT `native-linux-leds` child node provides a Linux LED name (as
+	  found under /sys/class/leds) via the 'path' property, and Zephyr LED
+	  brightness requests are mapped to /sys/class/leds/<path>/brightness.
diff --git a/drivers/led/linux_leds.c b/drivers/led/linux_leds.c
new file mode 100644
index 000000000000..5ebd93989ff7
--- /dev/null
+++ b/drivers/led/linux_leds.c
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2026 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_native_linux_leds
+
+#include <nsi_errno.h>
+#include <nsi_host_trampolines.h>
+#include <posix_native_task.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/fdtable.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(linux_leds, CONFIG_LED_LOG_LEVEL);
+
+struct linux_led_config {
+	const char *path;
+	int fd;
+	unsigned int max;
+};
+
+struct linux_leds_config {
+	struct linux_led_config *led;
+	int count;
+};
+
+static int linux_leds_set_brightness(const struct device *dev,
+				     uint32_t led_num, uint8_t value)
+{
+	const struct linux_leds_config *cfg = dev->config;
+	int val;
+	char buf[32];
+	int len;
+	int ret;
+
+	if (led_num >= cfg->count) {
+		return -EINVAL;
+	}
+
+	struct linux_led_config *led = &cfg->led[led_num];
+
+	/* round up so a binary led turns on any value > 0 */
+	val = DIV_ROUND_UP(value * led->max, LED_BRIGHTNESS_MAX);
+
+	len = snprintf(buf, sizeof(buf), "%d", val);
+	if (len < 0) {
+		return len;
+	}
+
+	ret = nsi_host_write(led->fd, buf, len);
+	if (ret != len) {
+		LOG_ERR("Write error: %s", strerror(nsi_host_get_errno()));
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static DEVICE_API(led, linux_leds_api) = {
+	.set_brightness = linux_leds_set_brightness,
+};
+
+static int linux_led_open(const char *path, int *max)
+{
+	/* 34: Length of /sys/class/leds/%s/max_brightness + \0 */
+	char full_path[strlen(path) + 34];
+	char buf[32];
+	int fd;
+	int ret;
+
+	ret = snprintf(full_path, sizeof(full_path), "/sys/class/leds/%s/max_brightness", path);
+	if (ret < 0) {
+		return ret;
+	} else if (ret >= sizeof(full_path)) {
+		return -ENOMEM;
+	}
+
+	fd = nsi_host_open(full_path, ZVFS_O_RDONLY);
+	if (fd < 0) {
+		LOG_ERR("Failed to open the led max brightness device %s: %s",
+			full_path, strerror(nsi_host_get_errno()));
+		return -EIO;
+	}
+
+	ret = nsi_host_read(fd, buf, sizeof(buf) - 1);
+	if (ret < 0) {
+		LOG_WRN("Read error: %s", strerror(nsi_host_get_errno()));
+		nsi_host_close(fd);
+		return -EIO;
+	}
+
+	buf[ret] = '\0';
+	*max = strtol(buf, NULL, 10);
+
+	nsi_host_close(fd);
+
+	ret = snprintf(full_path, sizeof(full_path), "/sys/class/leds/%s/brightness", path);
+	if (ret < 0) {
+		return ret;
+	} else if (ret >= sizeof(full_path)) {
+		return -ENOMEM;
+	}
+
+	fd = nsi_host_open(full_path, ZVFS_O_WRONLY);
+	if (fd < 0) {
+		LOG_ERR("Failed to open the led device %s: %s",
+			full_path, strerror(nsi_host_get_errno()));
+		return -EIO;
+	}
+
+	return fd;
+}
+
+static int linux_leds_init(const struct device *dev)
+{
+	const struct linux_leds_config *cfg = dev->config;
+	int ret;
+
+	for (int i = 0; i < cfg->count; i++) {
+		struct linux_led_config *led = &cfg->led[i];
+
+		ret = linux_led_open(led->path, &led->max);
+		if (ret < 0) {
+			goto cleanup;
+		}
+
+		led->fd = ret;
+	}
+
+	return 0;
+
+cleanup:
+	/* Close all fds we opened before failing */
+	for (int i = 0; i < cfg->count; i++) {
+		if (cfg->led[i].fd < 0) {
+			continue;
+		}
+
+		nsi_host_close(cfg->led[i].fd);
+		cfg->led[i].fd = -1;
+	}
+	return ret;
+}
+
+#define DEVICE_DT_GET_COMMA(node_id) DEVICE_DT_GET(node_id),
+
+static void linux_leds_cleanup(void)
+{
+	const struct device *devs[] = {
+		DT_FOREACH_STATUS_OKAY(DT_DRV_COMPAT, DEVICE_DT_GET_COMMA)
+	};
+	const int count = ARRAY_SIZE(devs);
+
+	for (int i = 0; i < count; i++) {
+		const struct linux_leds_config *cfg = devs[i]->config;
+
+		for (int j = 0; j < cfg->count; j++) {
+			if (cfg->led[j].fd < 0) {
+				continue;
+			}
+
+			nsi_host_close(cfg->led[j].fd);
+			cfg->led[j].fd = -1;
+		}
+	}
+}
+
+NATIVE_TASK(linux_leds_cleanup, ON_EXIT, 10);
+
+#define LED_CFG(n) {			\
+	.path = DT_PROP(n, path),	\
+	.fd = -1,			\
+}
+
+#define LINUX_LEDS_INIT(inst)							\
+	static const struct linux_leds_config linux_leds_cfg_##inst = {		\
+		.led = (struct linux_led_config []){				\
+			DT_INST_FOREACH_CHILD_SEP(inst, LED_CFG, (,))		\
+		},								\
+		.count = DT_INST_CHILD_NUM(inst),				\
+	};									\
+										\
+	DEVICE_DT_INST_DEFINE(inst, linux_leds_init, NULL,			\
+			      NULL, &linux_leds_cfg_##inst,			\
+			      POST_KERNEL, CONFIG_LED_INIT_PRIORITY,		\
+			      &linux_leds_api);
+
+DT_INST_FOREACH_STATUS_OKAY(LINUX_LEDS_INIT)
diff --git a/dts/bindings/led/zephyr,native-linux-leds.yaml b/dts/bindings/led/zephyr,native-linux-leds.yaml
new file mode 100644
index 000000000000..6ac7dc089a07
--- /dev/null
+++ b/dts/bindings/led/zephyr,native-linux-leds.yaml
@@ -0,0 +1,37 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 Google, LLC
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Linux LED backed LED device
+
+compatible: "zephyr,native-linux-leds"
+
+child-binding:
+  description: Linux LED child node
+
+  include:
+    - name: led-node.yaml
+      property-allowlist:
+        - label
+
+  properties:
+    path:
+      type: string
+      required: true
+
+examples:
+  - |
+    / {
+      leds {
+        compatible = "zephyr,native-linux-leds";
+
+        led0: led_0 {
+          /* /sys/class/leds/tpacpi::lid_logo_dot/brightness */
+          path = "tpacpi::lid_logo_dot";
+        };
+        led1: led_1 {
+          /* /sys/class/leds/tpacpi::power/brightness */
+          path = "tpacpi::power";
+        };
+      };
+    };
diff --git a/tests/drivers/build_all/led/app.overlay b/tests/drivers/build_all/led/app.overlay
index 3b5045378365..72b510f6a2d3 100644
--- a/tests/drivers/build_all/led/app.overlay
+++ b/tests/drivers/build_all/led/app.overlay
@@ -257,5 +257,17 @@
 			compatible = "leds-group-multicolor";
 			leds = <&test_pwm_led0>, <&test_pwm_led1>;
 		};
+
+		native_linux_leds {
+			compatible = "zephyr,native-linux-leds";
+
+			led_0 {
+				path = "dev::a";
+			};
+
+			led_1 {
+				path = "dev::b";
+			};
+		};
 	};
 };
